optimize file read in `Config::verify`

onur-ozkan opened the pull request:
`Config::verify` refactored to improve the efficiency and memory usage of file hashing.

the8472 commented on src/bootstrap/download.rs:
```diff
@@ -320,25 +320,42 @@ impl Config {
+
+        let file = t!(File::open(path));
+        let mut reader = BufReader::new(file);
+        let mut buffer = [0; 4096]; // read in chunks of 4KB
+
+        loop {
+            let read_len = t!(reader.read(&mut buffer));
```
You can avoid the extra copy by using `BufReaders` `BufRead::{consume, fill_buf}` methods.
Though the overhead is probably negligible.

onur-ozkan replied:
Isn't that what `read` already does?

the8472 replied:
Actually, just skip the buf-reader I guess if you're already providing a buffer.

onur-ozkan replied:
One of the main idea here is to iterate through the file content in small chunks of bytes, so that we don't load the entire buffer into memory all at once when reading it. So I have to provide limited(in size) buffer in each call on `BufRead::read`(which also does the `consume` from the reader automagically on each call).

the8472 replied:
`BufReader` already has an internal buffer. So one buffer should do.

onur-ozkan replied:
> BufReader already has an internal buffer. So one buffer should do.

I guess you mean `File` internal buffer? If so, I switched that one, the implementation relies on the host OS, but I guess it doesn't really matter that much.

the8472 replied:
No, `std::fs::File` does not have any userspace buffering.  `std::io::Bufreader` [does](https://doc.rust-lang.org/nightly/src/std/io/buffered/bufreader.rs.html#51)

onur-ozkan replied:
Ah I see the point, big thanks!

Mark-Simulacrum commented on src/bootstrap/download.rs:
```diff
@@ -320,25 +320,44 @@ impl Config {
+
+        let file = t!(File::open(path));
+        const BUFFER_SIZE: usize = 4096; // read in chunks of 4KB
+        let mut reader = BufReader::with_capacity(BUFFER_SIZE, file);
```
Why are we customizing the buffer size here vs. using the default in BufReader? If we *are* going to hard-code a buffer size, it seems like we might as well skip BufReader and just use `read(&mut buffer)` with a buffer we allocate here, which would also mean we can skip consume() since we wouldn't need any tracking of how much of that slice we've read.

onur-ozkan replied:
It was exactly like that(https://github.com/rust-lang/rust/pull/116635#discussion_r1356902119), I forgot removing the hard-coded size after moving to `fill_buf` and `consume`.

onur-ozkan committed:
optimize file read in Config::verify

`Config::verify` refactored to improve the efficiency and
memory usage of file hashing.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
add unit test for Config::verify

Signed-off-by: onur-ozkan <work@onurozkan.dev>

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit d16e89dc9d8ebfcf4a475585223fd23a5b36cee0 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum d16e89dc9d8ebfcf4a475585223fd23a5b36cee0 -->
<!-- homu: {"type":"Approved","sha":"d16e89dc9d8ebfcf4a475585223fd23a5b36cee0","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d16e89dc9d8ebfcf4a475585223fd23a5b36cee0 with merge dda7d4c9cf1112e98886df056b57e9f6ac3b6822...
<!-- homu: {"type":"BuildStarted","head_sha":"d16e89dc9d8ebfcf4a475585223fd23a5b36cee0","merge_sha":"dda7d4c9cf1112e98886df056b57e9f6ac3b6822"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6521649155/job/17711320615)
Approved by: Mark-Simulacrum
Pushing dda7d4c9cf1112e98886df056b57e9f6ac3b6822 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6521649155/job/17711320615"},"merge_sha":"dda7d4c9cf1112e98886df056b57e9f6ac3b6822"} -->

rust-timer commented:
Finished benchmarking commit (dda7d4c9cf1112e98886df056b57e9f6ac3b6822): [comparison URL](https://perf.rust-lang.org/compare.html?start=0d410be23c45e2f3567a6ec35985f690473f9176&end=dda7d4c9cf1112e98886df056b57e9f6ac3b6822&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0d410be23c45e2f3567a6ec35985f690473f9176&end=dda7d4c9cf1112e98886df056b57e9f6ac3b6822&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0d410be23c45e2f3567a6ec35985f690473f9176&end=dda7d4c9cf1112e98886df056b57e9f6ac3b6822&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.8%, 4.7%]   | 5     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-0.8%, -0.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0d410be23c45e2f3567a6ec35985f690473f9176&end=dda7d4c9cf1112e98886df056b57e9f6ac3b6822&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=0d410be23c45e2f3567a6ec35985f690473f9176&end=dda7d4c9cf1112e98886df056b57e9f6ac3b6822&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 624.729s -> 622.812s (-0.31%)
**Artifact size**: 305.51 MiB -> 305.53 MiB (0.01%)
